docs/PERSISTENCE.md

```diff
@@ -0,0 +1,57 @@
+
+## Operations
+The log stream provides few operations: create, append, read, get list of plogs, update the list of plogs (removing some plogs for GC). When the log stream creates a new plog or seals a plog, it should persist the create plog id and the sealed plog id and offset via `Metadata Manager`.
+- Create: The log stream receives no inputs. It will first create one plog: the active plog. Then, it will record the active plog id and mark this log stream to be ready to handle append and read operations.
+- Append offset 
+- Append without offset: This operation takes one input: `payload`. It will return the plog id to indicate which plog stores this payload, and the current offset of this plog. The log stream will first check whether the current plog will excess the size limit if it appends the payload to the current plog. If it will not exceed the size limit, the log stream client will append the input payload to the corresponding plog and return success. Otherwise, it will first seal the old plog, record the sealed offset of the old plog, create a new plog, record the plog id of the new plog, then write the payload to the new plog. We call this process `Plog Switch`. During this process, all the upcoming write should wait until the switching process finished in order to ensure correctness.
+- Read: This operation takes 3 inputs: `start plog id`, `start offset` and `size`. It is possible that the input size is too large so that the log stream will read the contents from multiple plogs. The log stream will first check how many plogs this read operation will involve by checking the vector of plogs it used. Then, it will read the requested contents from all the involved plogs asynchronously. Once it receives the response from all the plogs, it will return a vector of payloads.
+
+## Optimization
+When we switch the old plog to the new plog for the write operation, all the upcoming write requests will be blocked until the switching process finished, which is very inefficient. We used several optimization methods to improve the performance of this write operation:
+- Pre-allocated Plog: When we create the log stream, we will not only create one active plog. Instead, we will also create a pre-allocated next plog which is used to perform plog switch faster. Therefore, when we do the plog switch, we can use the pre-allocated plog as the new plog immediately, instead of waiting for creating a new plog. Then, we can create the new pre-allocated plog in the background afterward.
```

Comment:
This should probably be handled by the underlying PLOG library - it should keep a pre-allocated plog at hand to satisfy future calls immediately.

Reply:
Current implementation follows your description